utilities/src/with_std/dynamic_events.rs

```diff
@@ -0,0 +1,419 @@
+use core::ops::Deref;
+use scale_info::{
+	scale::{Compact, Decode},
+	PortableRegistry,
+};
+use scale_json::{
+	ext::{DecodeAsType, JsonValue},
+	Error as DecodeError, ScaleDecodedToJson,
+};
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+use thiserror::Error;
+
+/// The fully-scoped type name of the events that are stored in the System pallet.
+const SYSTEM_EVENTS_TYPE_NAME: &str = "frame_system::EventRecord";
+
+#[derive(Error, Debug, Clone)]
+pub enum DispatchError {
+	#[error("Dispatch error: {0}")]
+	DispatchError(JsonValue),
+	#[error("Module error ‘{name}‘ from pallet ‘{pallet}‘: ‘{error}‘")]
+	KnownModuleError { pallet: String, name: String, error: String },
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+/// A wrapper around a JSON value that represents a substrate EventRecord.
+pub struct DynamicEventRecord(JsonValue);
+
+impl Deref for DynamicEventRecord {
+	type Target = JsonValue;
+
+	fn deref(&self) -> &Self::Target {
+		&self.0
+	}
+}
+
+#[derive(Debug)]
+pub struct DynamicEvent(JsonValue);
+
+impl Deref for DynamicEvent {
+	type Target = JsonValue;
+
+	fn deref(&self) -> &Self::Target {
+		&self.0
+	}
+}
+
+impl DynamicEventRecord {
+	/// Wrap a JSON value that represents a substrate EventRecord.
+	pub fn new(json: JsonValue) -> Option<Self> {
+		if json.as_object().map_or(false, |obj| {
+			obj.contains_key("event") && obj.contains_key("phase") && obj.contains_key("topics")
+		}) {
+			Some(Self(json))
+		} else {
+			None
+		}
+	}
+
+	/// Returns the inner event if it matches the given extrinsic index.
+	pub fn extrinsic_event(&self, extrinsic_index: u64) -> Option<DynamicEvent> {
+		self.0
+			.pointer("/phase/ApplyExtrinsic")
+			.and_then(|index| index.as_u64())
+			.filter(|index| *index == extrinsic_index)
+			.map(|_| self.event())
+	}
+
+	/// Returns the inner event and it's extrinsic index if the event was triggered during the
+	/// ApplyExtrinsic phase.
+	pub fn indexed_extrinsic_event(&self) -> Option<(u64, DynamicEvent)> {
```

Comment:
nit: type alias, we possibly already have one for this iirc

```suggestion
	pub fn indexed_extrinsic_event(&self) -> Option<(ExtrinsicIndex, DynamicEvent)> {
```